coderd/httpmw/apikey.go

```diff
@@ -393,6 +393,23 @@ func ExtractAPIKey(rw http.ResponseWriter, r *http.Request, cfg ExtractAPIKeyCon
 		})
 	}
 
+	if roles.Status == database.UserStatusDormant {
+		// If coder confirms that the dormant user is valid, it can switch their account to active.
+		// nolint:gocritic
+		u, err := cfg.DB.UpdateUserStatus(dbauthz.AsSystemRestricted(ctx), database.UpdateUserStatusParams{
+			ID:        key.UserID,
+			Status:    database.UserStatusActive,
+			UpdatedAt: database.Now(),
+		})
+		if err != nil {
+			return write(http.StatusInternalServerError, codersdk.Response{
+				Message: internalErrorMessage,
+				Detail:  fmt.Sprintf("can't activate a dormant user: %s", err.Error()),
+			})
+		}
+		roles.Status = u.Status
+	}
+
```

Comment:
It seems cumbersome to have to call both `UpdateUserLastSeenAt` and `UpdateUserStatus`. Couldn't `UpdateUserLastSeenAt` handle the `dormant` -> `active` transition automatically?

Reply:
On the other hand `UpdateUserLastSeenAt` does not seem to have the main responsibility to change the user status. I think that I'd rather keep them separate.

But if we decide to extend `UpdateUserLastSeenAt` with "update user status" function, then we need to include also a condition to not touch "suspended" accounts. 

Let me know your thoughts!

Reply:
It should be a fairly simple case to add a `WHERE` clause to `UpdateUserLastSeenAt` to check the user status.

Reply:
`WHERE` clause might be not the correct one, but this could be:

```sql
UPDATE
    users
SET
    last_seen_at = $2,
    updated_at = $3,
    user_status = 
        CASE WHEN user_status = 'dormant'::user_status THEN 'active'::user_status
             ELSE user_status
        END
WHERE
    id = $1
RETURNING *;
```

Nevertheless, I'm still not convinced about its implications. I saw some references to:

`// Only update LastUsed once an hour to prevent database spam.` and I'd rather be precautious...

`// We only want to update this occasionally to reduce DB write
// load. We update alongside the UserLink and APIKey since it's
// easier on the DB to colocate writes.`

I guess that it is safer to keep it separate.

Reply:
That's fair. 👍